File: python/TwoPointers/ValidPalindrome.py
"""
125. Valid Palindrome
Difficulty: Easy
Topics: Two Pointers, String
"""
def isPalindrome(s: str) -> bool:
    left, right = 0, len(s) - 1

    while left < right:
        # whitespaces and other non alphanumeric characters are counted in the string 
        if not s[left].isalnum():
            left += 1
        elif not s[right].isalnum():
            right -= 1
        elif s[left].lower() != s[right].lower():
            return False
        else: 
            left += 1
            right -= 1

    return True

File: java/TwoPointers/ValidPalindrome.java
package java.TwoPointers;

public class ValidPalindrome {
    private static boolean isPalindromeHelper(char c) {
        return (c >= 'a' && c <= 'z') ||
        (c >= 'A' && c <= 'Z') ||
        (c >= '0' && c <= '9');
    }
    /*
     * 125. Valid Palindrome
     * Difficulty: Easy
     * Topics: Two Pointers, String
     */
    public static boolean isPalindrome(String s) {
        int left = 0; int right = s.length() - 1;

        while (left < right) {
            // whitespaces and other non alphanumeric characters are counted in the string 
            if (!isPalindromeHelper(s.charAt(left))) {
                left++;
            }
            else if (!isPalindromeHelper(s.charAt(right))) {
                right--;
            }
            else if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                return false;
            }
            else {
                left++;
                right--; 
            }
        }

        return true;

    }
}
